packages/core/src/services/chains/ChainStreamConsumer/index.ts

```diff
@@ -0,0 +1,174 @@
+import { MessageRole } from '@latitude-data/compiler'
+
+import {
+  ChainEvent,
+  ChainEventTypes,
+  ChainStepResponse,
+  RunErrorCodes,
+  StreamEventTypes,
+  StreamType,
+} from '../../../constants'
+import { objectToString } from '../../../helpers'
+import { AIReturn, Config } from '../../ai'
+import { ChainError } from '../ChainErrors'
+import { ValidatedStep } from '../ChainValidator'
+import { StreamConsumer } from './StreamConsumer'
+
+export function enqueueChainEvent(
+  controller: ReadableStreamDefaultController,
+  event: ChainEvent,
+) {
+  controller.enqueue(event)
+}
+
+function parseError(e: unknown) {
+  const isChainError = (e: unknown): e is ChainError<RunErrorCodes> =>
+    e instanceof ChainError
+
+  if (!isChainError(e)) {
+    return new ChainError({
+      code: RunErrorCodes.Unknown,
+      message: (e as Error).message,
+    })
+  }
+
+  return e
+}
+
+export class ChainStreamConsumer {
+  private controller: ReadableStreamDefaultController
+  private previousCount: number
+  private documentLogUuid: string
+  private streamConsumer: InstanceType<typeof StreamConsumer>
+
+  static parseResponseText(response: ChainStepResponse<StreamType>) {
+    const isStreamText = response.streamType === 'text'
+    const isStreamObject = response.streamType === 'object'
+    return isStreamText
+      ? response.text
+      : isStreamObject
+        ? objectToString(response.object)
+        : ''
```

Comment:
hmm is this new? why are we defaulting to empty string if there is no response? it should be undefined which is not the same than empty string

Reply:
I'll review this one

Reply:
I think we already were doing this on chain completed
![image](https://github.com/user-attachments/assets/3d999f5d-2751-47c3-809c-3d1f322b2318)